src/fpm/manifest.f90

```diff
@@ -103,13 +104,17 @@ subroutine get_package_data(package, file, error, apply_defaults)
         !> Apply package defaults (uses file system operations)
         logical, intent(in), optional :: apply_defaults
 
+        !> Add `FPM_IS_WINDOWS` macro to the preprocessor
+        logical, intent(in), optional :: add_is_windows_macro
+
         type(toml_table), allocatable :: table
         character(len=:), allocatable :: root
+        logical :: set_is_windows_macro = .true.
```

Comment:
Please help me understand here. Are we adding the `FPM_IS_WINDOWS` macro to *every* package that fpm is going to ever build?

Reply:
In the beginning, I added it to the root package only, but it didn't work (it wasn't registered during compilation). Maybe the reason is that the root package is also added to the dependency tree and that's the place where the macro matters. It worked when I exported it for every dependency in the dependency tree. I did so also because I didn't want to differentiate between the root package and all the other dependencies in the dependency tree.

If you have any better idea, let me know. I'll see if I can add the dependency to the dependency tree with the flag in place.

Reply:
Yes, I had a look again. `get_package_data` is called on every package of the dependency tree (again) and that's where the macro needs to be set for the root package. So if I did this, it works:

```fortran
if (i == 1) then
    call get_package_data(dependency, manifest, error, apply_defaults=.true., add_is_windows_macro=.true.)
else
    call get_package_data(dependency, manifest, error, apply_defaults=.true., add_is_windows_macro=.false.)
end if
```

So I'm open for better suggestions.

Reply:
What we could do is add an opt-in instead of an opt-out. Meaning that we could add an `export-windows-macro` flag to the `[build]` config, for example, and always set it to `true` in the `fpm` manifest.

Reply:
My understanding is that minGW/MSYS compilers don't have access to `realpath` on Windows. So I would start looking for a kiss fix first. What would happen if we just extended this for macros `_WIN64`,  `__MINGW32__`, `__MSYS__`, `__MINGW64__`: 

https://github.com/fortran-lang/fpm/blob/fc4ac0e349a30460ecc07fa7b47473e0be19e25c/src/fpm_os.c#L11-L15

Reply:
Do minGW/MSYS compilers have `fullpath`?

Reply:
That part of the code has so far not been a problem, but the bootstrapping is. If that part of the code you're referring to needs a fix, I'm happy to open another PR in the future.

Reply:
@minhqdao @zoziha what please take a look at this approach:

https://github.com/fortran-lang/fpm/compare/main...perazz:fpm:test_windows_macros

I can't test it on msys2 right now (the CI has minGW64) but I believe it should work.

Reply:
@perazz The `gfortran` compiler only presets the macro definition `__GFORTRAN__`, other macro definitions such as `_WIN32`, `_WIN64`, `__MINGW64__` are not preset.

Reply:
What gfortran provider on Windows has this issue? I understood it is from MSYS, or is it another installation? If it is minGW or MSYS, then some of those macros should be defined I believe.

Reply:
The previous Fortran standard did not specify preprocessors, and Fortran should not support preprocessors, but compiler manufacturers still adapted C preprocessors for Fortran. GCC (C/C++ language compiler) has these preset macro definitions, while gfortran does not, and I haven't seen any gfortran providers (msys, cygwin, mingw) on Windows preset macro definition _WIN32.